SQL Lookup Tables - Part 1

millerdev opened the pull request:
This is the [first PR](https://commcare-hq.readthedocs.io/couch_to_sql_models.html#pr-1-add-sql-model-and-migration-management-command-write-to-sql) in the migration of Lookup Tables (previously known as Fixtures) from Couch to SQL.

Lookup Table performance has been an issue in the past. In general new SQL operations should perform better than the equivalent Couch operations. Bulk operations have been added where the code was previously saving row-by-row. Overall write performance will obviously be at least slightly worse until we can stop writing to Couch (a couple of PRs away). Specifically, performance of the Excel uploader and web interface for editing tables will be impacted while writing to Couch, but other read-only operations such as restores are not expected to be affected.

~A change log entry with instructions for how to migrate external environments will be added to commcare-cloud after this has been merged.~

:tropical_fish: Review by commit.

The full set of changes, including commits to be included in future PRs can be found at https://github.com/dimagi/commcare-hq/compare/dm/sql-lookup-tables - I plan to rebase that branch on this and future PRs as modifications are made, FYI.

## Safety Assurance

### Safety story

Test driven development was used to ensure existing features continue to function with SQL as they did before on Couch.

### Automated test coverage

Yes. New tests have been added in places where they were previously missing.

### QA Plan

Lookup Table features to be tested in QA:

- Excel uploads.
- Excel downloads.
- Updates using the web interface.
- APIs: read and write operations.
- Linked domains: Lookup Table changes in the parent domain should be propagated to child domains.
- Restores containing lookup tables, including
  - global tables
  - tables with user-, group-, and location-owned rows.
- Other Lookup Table features in HQ?

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-4392).

### Migrations
- [x] The migrations in this code can be safely applied first independently of the code.

### Rollback instructions

The migration in this PR adds new SQL tables and indexes. The presence of those entities does not pose a problem for rollback, but subsequent PRs must account for the fact that the migration has been applied. It may be desirable to unapply the migration in a private release after rollback to allow a fresh start.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

millerdev committed:
Lint

millerdev committed:
Add 'multi_populate' management command

Intended for running multiple subclasses of PopulateSQLCommand in series.

Example usage:
./manage.py multi_populate populate_lookuptables populate_lookuptablerows populate_lookuptablerowowners

millerdev committed:
Make CouchTransaction sync Couch changes to SQL

millerdev committed:
Add jsonattrs

A Pythonic way to interact with structured JSON in Django models.

millerdev committed:
Use transaction for SQL Lookup Table upload

millerdev committed:
Migrate FixtureDataType to LookupTable

millerdev committed:
Handle old FixtureDataType formats in migration

FixtureDataType.wrap() handles these lazily, but the migration does not use that.

millerdev committed:
Adapt "update_lookup_tables" view to write to SQL

millerdev committed:
Add CharIdField to avoid varchar_pattern_ops indexes

millerdev committed:
Migrate FixtureDataItem to LookupTableRow

millerdev committed:
Migrate FixtureOwnership to LookupTableRowOwner

snopoke commented on corehq/apps/fixtures/views.py:
```diff
@@ -226,6 +233,30 @@ def _update_items(fields_patches, domain, data_type_id, transaction):
+            )
+        elif "remove" in patch:
+            fields_json = JsonDelete(fields_json, field_name)
+    for field_name, patch in fields_patches.items():
```
why does this have to be in a separate loop?

millerdev replied:
This is how the code was structured before, and it was hard to evaluate whether it was strictly necessary. The code in this area looks fragile and possibly buggy to me, but fixing those issues is out of scope for this PR.

snopoke commented on corehq/sql_db/jsonops.py:
```diff
@@ -0,0 +1,69 @@
+        the value in `expression` to be replaced.
+    :param new_value: JSON-serializable value or expression.
+    :param create_missing: Add value if it is missing (default: True).
+    """
```
a link to the PG docs or else an example of what this renders to would be useful.

millerdev replied:
59b3dcd2fb561014a8b7c8625353d0ffb3922ee1

snopoke commented on corehq/sql_db/jsonops.py:
```diff
@@ -0,0 +1,69 @@
+import json
```
could do with a module docstring illustrating usage, particularly how they can be nested

snopoke replied:
also, how about some tests?

millerdev replied:
59b3dcd2fb561014a8b7c8625353d0ffb3922ee1

snopoke commented on corehq/apps/fixtures/views.py:
```diff
@@ -226,6 +233,30 @@ def _update_items(fields_patches, domain, data_type_id, transaction):
         setattr(item, "fields", updated_fields)
         transaction.save(item)
+
+    fields_json = "fields"
+    for field_name, patch in fields_patches.items():
+        if "update" in patch:
+            new_field_name = patch["update"]
+            fields_json = JsonSet(
+                JsonDelete(fields_json, field_name),
+                [new_field_name],
+                JsonGet("fields", field_name),
+            )
+        elif "remove" in patch:
+            fields_json = JsonDelete(fields_json, field_name)
+    for field_name, patch in fields_patches.items():
+        if "is_new" in patch:
+            fields_json = JsonSet(fields_json, [field_name], [])
```
phew! took me a bit to get my head around this whole nested json editing bit. Great that we can do a bulk update!

snopoke commented on corehq/apps/fixtures/apps.py:
```diff
@@ -0,0 +1,6 @@
+
+class Config(AppConfig):
+    name = 'corehq.apps.fixtures'
+    default_auto_field = 'django.db.models.BigAutoField'
```
seems like this isn't actually relevant to the models since they are either using `AutoField` or `UUID`

millerdev replied:
It's used by the implicit `id` of `LookupTableRowOwner`. I also want any new models in this app to default to this if not otherwise specified.

snopoke commented on corehq/apps/fixtures/models.py:
```diff
@@ -14,6 +22,231 @@
+
+
+@unique
+class OwnerType(IntEnum):
```
did you consider using `django.models.IntegerChoices` (https://docs.djangoproject.com/en/4.0/ref/models/fields/#enumeration-types)? You would get the unique and `choices` method for free. You'd still have to add `from_string`

millerdev replied:
I think I wrote this before we had migrated to Django 3. I intended to take another look, but forgot about it so thanks for the reminder. 7852cb6996e2d834829f046ed9e565025efe786c

millerdev commented on corehq/apps/fixtures/couchmodels.py:
```diff
@@ -83,6 +100,32 @@ def wrap(cls, obj):
+            "domain",
+            "is_global",
+            "tag",
+            "item_attributes",
```
It's not used at all.

> If not what about removing it from the model?

Do you mean remove it from the Couch model? All of the Couch models will eventually be removed (in the 3rd PR in this series), so it felt out of scope for this PR.

millerdev commented on corehq/apps/fixtures/models.py:
```diff
@@ -111,6 +111,97 @@ def _migration_get_or_create_couch_object(self):
+    class Meta:
+        app_label = 'fixtures'
+        indexes = [
+            models.Index(fields=["domain", "table_id", "sort_key", "id"]),
```
All these fields are necessary for paginated iteration over rows ordered by `sort_key`. The manager method to do that will be added in the next PR.

https://github.com/dimagi/commcare-hq/blob/97e4cd7533707c1f4b0d26ae0290cfa9b1fe7139/corehq/apps/fixtures/models.py#L180-L196

millerdev committed:
Use OwnerType.from_string instead of dict constant

millerdev committed:
Use contextlib.closing instead of try/finally

millerdev committed:
Improve docs and add tests for jsonops

millerdev commented on corehq/util/jsonattrs.py:
```diff
@@ -0,0 +1,235 @@
+"""A Pythonic way to interact with structured JSON in Django models.
```
Would plain attrs classes and `attrs.asdict` do most of what you're thinking of? Maybe we can make `dict_of` and `list_of` more useful in stand-alone contexts? I'm inclined to do that in a separate PR where there is a use case for it.

snopoke replied:
I think the main use case is going from JSON to Python (with the support for nested objects).

millerdev committed:
IntegerChoices instead of IntEnum with @unique

millerdev committed:
Fix test_deletion_sql_models failure

millerdev commented on corehq/apps/fixtures/management/commands/populate_lookuptablerowowners.py:
```diff
@@ -0,0 +1,66 @@
+        return model, created
+
+
+OWNER_TYPE_MAP = {
```
be130bb7eb5144ca35a2349087e1ccc5e168975a

millerdev commented on corehq/util/test_utils.py:
```diff
@@ -875,6 +876,25 @@ def print_sql_chronologically(self, with_traceback=False, width=150):
+        with mock.patch("django.db.connections._connections.default", cn):
+            yield cn
+    finally:
+        cn.close()
```
239d1442da48393e7005b3262410bce4cee0686c

millerdev committed:
Change wrapping to appease stickler

The previous style works on Python 3.9, but maybe stickler is running an older version?

dannyroberts commented on corehq/sql_db/tests/test_jsonops.py:
```diff
@@ -0,0 +1,139 @@
+        .annotate(val=expression)
+        .values_list("val", flat=True)
+    )
+    print(qs.query)
```
Is this a debug print you meant to remove or useful test output?

millerdev replied:
Useful test output.

snopoke commented on corehq/sql_db/jsonops.py:
````diff
@@ -1,3 +1,31 @@
+# Before: {"values": [1, 2, 3], "other": "something", "score": 3}
+# After:  {"items": [1, 2, 3], "other": "something"}
+```
+"""
````
👍🏻
